specs/AppLifecycle/Restart/restartApi.markdown

```diff
@@ -0,0 +1,254 @@
+<li><p>Kind: Gets the reason that this app is being activated. This is of type enum ActivationKind</p></li>
+</ul>
+<p>PreviousExecutionState: Gets the execution state of the app before this activation.</p></td>
+</tr>
+</tbody>
+</table>
+
+## New API Details
+
+### RequestRestartNow
+
+<table>
+<colgroup>
+<col style="width: 41%" />
+<col style="width: 53%" />
+<col style="width: 5%" />
+</colgroup>
+<thead>
+<tr class="header">
+<th>New API </th>
+<th>Description</th>
+</tr>
+</thead>
+<tbody>
+<tr class="odd">
+<td>static AppRestartFailureReason <strong>RequestRestartNow</strong>(string restartArguments)</td>
+<td><p>Equivalent to CoreApplication.RequestRestartAsync, except that it is <em>not</em> async.
+<ul>
+<li>If the restart fails, but the user subsequently launches the app manually, the app will launch normally and no restart arguments will be passed. </li>
+<li>If the app has any in-process background tasks running when it calls this API, those tasks will be cancelled in the normal way. Out-of-process background tasks will not be affected. </li>
+</ul>
+ The restarted instance will be activated with Windows.ApplicationModel.Activation.LaunchActivatedEventArgs.</p>
+ <ul>
+<li>When the app is restarted, LaunchActivatedEventArgs.PreviousExecutionState will have the value Terminated so that the app can distinguish between a resume and a restart. </li>
+</ul>
+<p>Returns a failure reason for why the restart didn’t successfully complete (AppRestartFailureReason).
+<ul>
+<li>This API leverages an existing Enum (AppRestartFailureReason), but not all return values are relevant.</li>
+<li>This API does not require for the app to be in foreground, and so AppRestartFailureReason::NotInForeground will never be returned</li>
+</ul>
+</p></td>
+</tr>
+</tbody>
+</table>
+
+When an application calls RequestRestartNow, the mechanism for restart will be as follows:
+1. Application calls RequestRestartNow API.
+2. The API will CreateProcess with the Agent (a helper EXE in the framework package containing the API) and will then wait for an an acknowledgment (ACK) from the Agent.
+3. The Agent will get the executable path of the application and signal an ACK to the API. The Agent will then then terminate the application and then CreateProcess with the application (restarted).
+
+## Examples and Scenarios
+
+```c#
+public class App 
+{
+    // Let's assume this method is run when an app updates assets.
+    // The new assets have downloaded and installed, and now the app has to request a restart immediately
+    protected void UpdateInstallComplete()
+    {
+        // Checking if the assets have updated successfully and there are no pending updates
+        if (!CheckForUpdate())
+        {
+            AppRestartFailureReason result = AppLifecycle::RequestRestartNow("args for new instance");
+            switch (result)
+            {
+                case AppRestartFailureReason::RestartPending:
```

Comment:
RestartPending is only returned if another restart is in progress.  Since that is more likely for the async variation in the platform, this should only ever happen if the API is called twice by the same process.

Reply:
Updated with this information.